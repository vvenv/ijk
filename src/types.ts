export interface TemplateOptions {
  debug?: boolean;
  strictMode?: boolean;
  autoEscape?: boolean;
  collapseWhitespace?: boolean;
  stripComments?: boolean;
  tagStart?: string;
  tagEnd?: string;
}
