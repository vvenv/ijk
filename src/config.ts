export const CONTEXT = 'c';
export const FILTERS = 'f';
export const ESCAPE = 'e';
export const BLOCK = 'block';
export const TAG_START = '{{';
export const TAG_END = '}}';
export const ROOT = 'root';
