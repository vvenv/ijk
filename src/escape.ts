export const escape = (v: unknown) => {
  return `${v}`
    .replace(/&/g, '&amp;')
    .replace(/"/g, '&#34;') // as it may be used in attributes
    .replace(/'/g, '&#39;'); // as it may be used in attributes
};
