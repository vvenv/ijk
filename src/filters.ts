import { Safe } from './safe';

export const abs = (value: number) => Math.abs(value);
export const capitalize = (value: string) =>
  value.charAt(0).toUpperCase() + value.slice(1);
export const date = (value: string) => new Date(value).toISOString();
export const entries = (value: object) => Object.entries(value);
export const escape = (value: string) =>
  value.replace(/</g, '&lt;').replace(/>/g, '&gt;');
export const even = (value: number | string) => +value % 2 === 0;
export const first = (value: string | any[]) => [...value][0];
export const join = (value: string[], separator = '') => value.join(separator);
export const json = (value: any) => JSON.stringify(value);
export const keys = (value: object) => Object.keys(value);
export const last = (value: string | any[]) => [...value].reverse()[0];
export const length = (value: string) => value.length;
export const lower = (value: string) => value.toLowerCase();
export const odd = (value: number | string) => +value % 2 === 1;
export const repeat = (value: string, count: number) => value.repeat(count);
export const replace = (value: string, search: string, replace: string) =>
  value.replace(new RegExp(search, 'g'), replace);
export const reverse = (value: string | any[]) =>
  Array.isArray(value) ? value.reverse() : [...value].reverse().join('');
export const safe = (value: string) => new Safe(value);
export const slice = (value: string, start: number, end?: number) =>
  value.slice(start, end);
export const sort = (value: string | any[]) =>
  Array.isArray(value) ? value.sort() : [...value].sort().join('');
export const split = (value: string, separator = '') => value.split(separator);
export const sum = (value: number[]) => value.reduce((a, b) => a + b, 0);
export const time = (value: string) => new Date(value).getTime();
export const trim = (value: string) => value.trim();
export const unique = (value: string | any[]) =>
  Array.isArray(value)
    ? Array.from(new Set(value))
    : Array.from(new Set(value)).join('');
export const upper = (value: string) => value.toUpperCase();
export const values = (value: object) => Object.values(value);
