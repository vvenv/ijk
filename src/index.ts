import { Template } from './template';
import { TemplateOptions } from './types';

export * from './template';

export const template = (
  template: string,
  data: object,
  options?: TemplateOptions,
) => new Template(options).compile(template).render(data);
