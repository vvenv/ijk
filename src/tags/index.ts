export * from './assign';
export * from './block';
export * from './comment';
export * from './for';
export * from './if';
export * from './macro';
export * from './variable';
