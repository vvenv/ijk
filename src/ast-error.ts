import { bgRed, red, bold, gray, $ } from 'kleur/colors';
import { AST, EndTag, StartTag } from './ast';

interface ASTErrorOptions {
  ast: AST;
  tags: Partial<StartTag | EndTag>[];
}

$.enabled = typeof process === 'undefined' || process.env.NODE_ENV !== 'test';

export class ASTError extends SyntaxError {
  constructor(
    message: string,
    private options: ASTErrorOptions,
  ) {
    super(message);
    this.name = 'ASTError';
    Error.captureStackTrace?.(this, this.constructor);
  }

  get details() {
    const output: string[] = [];
    const caretLines = new Set<number>();

    const addCaretLine = (index: number) => {
      caretLines.add(index - 2);
      caretLines.add(index - 1);
      caretLines.add(index);
      caretLines.add(index + 1);
      caretLines.add(index + 2);
    };

    const lines = this.options.ast.template.split('\n');
    const indentWidth = String(lines.length).length + 2;
    const tags = [...this.options.tags];

    let cursor = 0;
    let caretLinesCount = 0;

    lines.forEach((line, index) => {
      output.push(
        `${gray(`${`${index + 1}: `.padStart(indentWidth, ' ')}`)}${line}`,
      );

      tags.forEach((tag) => {
        if (tag.startIndex! < cursor) {
          tags.splice(tags.indexOf(tag), 1);
          return;
        }
        if (tag.startIndex! >= cursor + line.length + 1) {
          return;
        }

        const offset = tag.startIndex! - cursor + indentWidth;
        const end = Math.min(tag.endIndex!, cursor + line.length);
        const count = end - tag.startIndex!;

        output.push(`${' '.repeat(offset)}${red('^'.repeat(count))}`);

        if (end < tag.endIndex!) {
          tag.startIndex! += count + 1;
        }

        addCaretLine(++caretLinesCount + index);
      });

      cursor += line.length + 1;
    });

    const emptyLine = `${gray(`${' '.repeat(indentWidth)}...`)}`;

    return `${bgRed(bold(' i>j>k '))} ${red(this.message)}

${output
  .map((line, index) => (caretLines.has(index) ? line : emptyLine))
  .reduce(
    (acc, line) =>
      line === emptyLine && acc.at(-1) === line ? acc : [...acc, line],
    [] as string[],
  )
  .join('\n')}
`;
  }
}
