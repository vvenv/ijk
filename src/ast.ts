import { ASTError } from './ast-error';
import { TemplateOptions } from './types';

export interface EndTag {
  name: string;
  startIndex: number;
  endIndex: number;
  block: ASTNode;
  prev: StartTag;
  next: null;
}

export interface StartTag {
  name: string;
  statement?: string;
  startIndex: number;
  endIndex: number;
  node: ASTNode;
  prev: StartTag | null;
  next: StartTag | EndTag | null;
  children: ASTNode[];
}

export interface ASTNode {
  tags: [...StartTag[], EndTag];
  parent: ASTNode | null;
  prev: ASTNode | null;
  next: ASTNode | null;
  level: number;
  index: number;
  skip?: boolean;
}

export class AST implements ASTNode {
  private cursor: ASTNode;

  tags: [StartTag, EndTag];
  parent: null;
  prev: null;
  next: null;
  level: number;
  index: number;

  constructor(
    public template = '',
    public options: Required<TemplateOptions>,
  ) {
    this.tags = [] as unknown as [StartTag, EndTag];
    this.parent = null;
    this.prev = null;
    this.next = null;
    this.level = 0;
    this.index = 0;

    this.cursor = this;
  }

  get valid() {
    return this.cursor === this && this.tags.length % 2 === 0;
  }

  get children() {
    return this.tags[0]?.children ?? [];
  }

  start(tag: Partial<StartTag>): StartTag {
    const { tags } = this.cursor;

    const startTag = {
      ...tag,
      prev: null,
      next: null,
      children: [],
    } as StartTag;

    // []
    if (tags.length === 0) {
      startTag.node = this.cursor;
      tags.push(startTag);
      return startTag;
    }

    // [...StartTag]
    const lastTag = tags.at(-1) as StartTag;
    const lastBlock = lastTag.children.at(-1) ?? null;

    const node: ASTNode = {
      tags: [startTag] as unknown as [StartTag, EndTag],
      parent: this.cursor,
      prev: lastBlock ?? null,
      next: null,
      level: this.cursor.level + 1,
      index: lastTag.children.length,
    };

    if (lastBlock) {
      lastBlock.next = node;
    }
    startTag.node = node;
    lastTag.children.push(node);
    this.cursor = node as ASTNode;

    return startTag;
  }

  between(tag: Partial<StartTag>) {
    const { tags } = this.cursor;

    if (!tags.length) {
      if (this.options.debug) {
        throw new ASTError(`Unexpected ${tag.name} tag`, {
          ast: this,
          tags: [tag],
        });
      }

      return;
    }

    const lastTag = tags.at(-1) as StartTag;
    const _tag = {
      ...tag,
      node: lastTag.node,
      prev: lastTag,
      next: null,
      children: [],
    } as StartTag;

    lastTag.next = _tag;
    tags.push(_tag);
  }

  end(tag: Partial<EndTag>) {
    const { tags } = this.cursor;

    if (!tags.length) {
      if (this.options.debug) {
        throw new ASTError(`Unexpected ${tag.name} tag`, {
          ast: this,
          tags: [tag],
        });
      }

      return;
    }

    const [startTag] = tags as [StartTag, EndTag];

    const expected = `end${startTag.name}`;
    if (expected !== tag.name) {
      if (this.options.debug) {
        console.trace(tag, tags);
        throw new ASTError(
          `Unexpected "${tag.name}" tag, expected "${expected}"`,
          {
            ast: this,
            tags: [tag],
          },
        );
      }

      return;
    }

    const lastTag = tags.at(-1) as StartTag;
    const _tag = {
      ...tag,
      block: startTag.node,
      prev: lastTag,
      next: null,
    } as EndTag;

    lastTag.next = _tag;
    tags.push(_tag);

    // It's a close block, so we need to move the cursor to the parent
    this.cursor = this.cursor.parent ?? this;
  }

  private throw(message: string, ...tags: Partial<StartTag | EndTag>[]) {
    if (this.options.debug) {
      throw new ASTError(message, {
        ast: this,
        tags,
      });
    }
    return false;
  }

  getFirstTag() {
    return this.cursor.tags.at(0);
  }

  assertFirstTag(
    name: string,
    tag: Partial<StartTag | EndTag>,
    required = true,
  ) {
    const firstTag = this.cursor.tags.at(0)!;
    if (firstTag.name !== name) {
      if (required) {
        this.throw(
          `"${tag.name}" must follow "${name}", not "${firstTag.name}".`,
          firstTag,
          tag,
        );
      }
      return false;
    }
    return true;
  }

  getLastTag() {
    return this.cursor.tags.at(-1);
  }

  getNextTag(tag: StartTag) {
    return tag.children[0]?.tags[0] ?? tag.next;
  }

  getPrevTag(tag: StartTag) {
    return tag.prev ?? tag.node.prev?.tags.at(-1);
  }
}
