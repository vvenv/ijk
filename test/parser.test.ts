import { describe, expect, test } from 'vitest';
import { parse } from './__helper';
import { CommentTag, ForTag } from '../src/tags';
import { TemplateOptions } from '../src/types';

const _parse = (input: string, options?: TemplateOptions) =>
  parse(input, options, [ForTag, CommentTag]);

describe('collapseWhitespace', () => {
  test('on', () => {
    expect(
      _parse(
        `{{ for name in names }}
  {{ name }} in {{ names }}
{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });

  test('off', () => {
    expect(
      _parse(
        ` {{ for name  in names }}
  {{ name }} in  {{ names }}
 {{ endfor  }} `,
        {
          collapseWhitespace: false,
        },
      ),
    ).toMatchSnapshot();
  });
});

describe('stripComments', () => {
  test('on', () => {
    expect(_parse(`{{ ! this is a comment }}`)).toMatchSnapshot();
  });

  test('off', () => {
    expect(
      _parse(`{{ ! this is a comment }}`, {
        stripComments: false,
      }),
    ).toMatchSnapshot();
  });
});

describe('strictMode', () => {
  test('on', () => {
    expect(_parse(``)).toMatchSnapshot();
  });
  test('off', () => {
    expect(
      _parse(``, {
        strictMode: false,
      }),
    ).toMatchSnapshot();
  });
});

test('tagStart and tagEnd', () => {
  expect(
    _parse(`{% for name in names %}{{ name }} in {{ names }}{% endfor %}`, {
      tagStart: '{%',
      tagEnd: '%}',
    }),
  ).toMatchSnapshot();
});

test('empty', () => {
  expect(_parse(``)).toMatchSnapshot();
});

test('html tags', () => {
  expect(_parse(`<foo>foo</foo>`)).toMatchSnapshot();
});

test('quotes', () => {
  expect(_parse(`"'foo'"`)).toMatchSnapshot();
});

test('line break feed', () => {
  expect(_parse(`\nfoo\n`)).toMatchSnapshot();
});
