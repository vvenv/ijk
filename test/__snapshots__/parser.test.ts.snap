// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`collapseWhitespace > off 1`] = `""use strict";let s="";s+=" ";const o_1_0=c.names;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+="\\n  ";s+=e(c_i_1_0.name);s+=" in  ";s+=e(c_i_1_0.names);s+="\\n ";}s+=" ";return s;"`;

exports[`collapseWhitespace > on 1`] = `""use strict";let s="";const o_1_0=c.names;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=" ";s+=e(c_i_1_0.name);s+=" in ";s+=e(c_i_1_0.names);s+=" ";}return s;"`;

exports[`empty 1`] = `""use strict";let s="";return s;"`;

exports[`html tags 1`] = `""use strict";let s="";s+="<foo>foo</foo>";return s;"`;

exports[`line break feed 1`] = `""use strict";let s="";s+=" foo ";return s;"`;

exports[`quotes 1`] = `""use strict";let s="";s+="\\"'foo'\\"";return s;"`;

exports[`strictMode > off 1`] = `"let s="";return s;"`;

exports[`strictMode > on 1`] = `""use strict";let s="";return s;"`;

exports[`stripComments > off 1`] = `""use strict";let s="";s+="<!--this is a comment-->";return s;"`;

exports[`stripComments > on 1`] = `""use strict";let s="";return s;"`;

exports[`tagStart and tagEnd 1`] = `""use strict";let s="";const o_1_0=c.names;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+="{{ name }} in {{ names }}";}return s;"`;
